### ISSUE TITLE
[Feature]: Automatically set `process.env.PLAYWRIGHT = 'true'`?

### ISSUE THREAD

**Issue by karlhorky** [microsoft/playwright#32356]
### 🚀 Feature Request

Similar to [Vitest setting `process.env.VITEST = 'true'` by default](https://github.com/vitest-dev/vitest/blob/122e4fe0a7d4f208205617069b3df6a0ffaf7467/packages/vitest/src/node/pool.ts#L124), what does the Playwright team think about setting `process.env.PLAYWRIGHT` to true?

### Example

For tests of features which do not work in Playwright (eg. reCAPTCHA or other connections to external services), disabling them via `process.env.PLAYWRIGHT` can be useful

### Motivation

Described above in Example

**Comment by Skn0tt** [comment]
The usecase sounds great, but I think the alternative solution you mention is pretty fine already. One downside to the automatically-set env var is that folks will expect it not just in their web server, but also in their client-side code that might be bundled separately from the webserver. So i'm wary of introducing too much magic here, especially since Playwright doesn't control enough of the build process to properly do so.

Could you help me understand how an automatically-set `PLAYWRIGHT` env var would be better than the alternative you mention?

**Comment by karlhorky** [comment]
> One downside to the automatically-set env var is that folks will expect it not just in their web server, but also in their client-side code that might be bundled separately from the webserver. So i'm wary of introducing too much magic here, especially since Playwright doesn't control enough of the build process to properly do so.

Good point, I'm running into this actually already now 😅

I'm considering what the cleanest approach to injecting this in the build process and the Playwright web servers is...

Currently (in addition to the extra configuration in `playwright.config.ts`), in my `package.json` I have something like this:

`package.json`

```json
{
  "scripts": {
    "test": "PLAYWRIGHT=true pnpm build-1 && pnpm build-2 && yarn fusion playwright test --ui"
  }
}
```

However, `PLAYWRIGHT='true'` doesn't make its way to the `build-2` step 🤔 I'm guessing this is just a detail of Bash execution or npm scripts that I'm missing, but I don't want to add more duplication (or complexity / dependencies) to propagate this environment variable.

Then also, to pass this along to client-side React code in Next.js, adding this again in the `env` configuration in `next.config.ts`:

`next.config.ts`

```ts
import { NextConfig } from 'next';

const config: NextConfig = {
  env: {
    PLAYWRIGHT: process.env.PLAYWRIGHT,
  },
```

> Could you help me understand how an automatically-set `PLAYWRIGHT` env var would be better than the alternative you mention?

Actually for a the reason that you mentioned and my further elaboration above - to reduce the duplicated, scattered configuration across multiple files for propagating this `PLAYWRIGHT` environment variable.

It wouldn't completely remove the need for some configuration (see the `next.config.ts` example above), but it would go in the direction of zero-configuration at least.

> since Playwright doesn't control enough of the build process to properly do so

Would Playwright be open to accepting + running a `webServer[number].buildCommand` string to first run a build process before a webServer start? Both the `webServer[number].command` and the `webServer[number].buildCommand` could be run with `PLAYWRIGHT='true'` in this case, allowing for lower-configuration environment variable propagation.

**Comment by Skn0tt** [comment]
Instead of `buildCommand`, have you tried adding another webServer that just contains the build command, but no `url` to wait for? From my understanding of the [code](https://github.com/microsoft/playwright/blob/6763d5ab6bd20f1f0fc879537855a26c7644a496/packages/playwright/src/plugins/webServerPlugin.ts#L83), that should do exactly what you're looking for. We don't seem to have a test for that behaviour, but I can add that.

**Comment by karlhorky** [comment]
> Instead of `buildCommand`, have you tried adding another webServer that just contains the build command, but no `url` to wait for?

hmm... do you mean to add build commands in new `webServer` array objects before every web server start object?

I assume that implies that the `webServer` array is run sequentially from top to bottom? Order would be important here - eg. it would be problematic if the start command in `webServer[number].command` ran before the build completed...

Would each "build" `webServer` object come directly before each "start" `webServer` object? (option 1: grouping by service) Or would all build objects come at the start of the array? (option 2: grouping by stage)

Here's an example of option 1: grouping by service:

`playwright.config.ts`

```ts
import { devices, PlaywrightTestConfig } from '@playwright/test';

const config: PlaywrightTestConfig = {
  webServer: [
    {
      command: 'pnpm website build',
      env: {
        PLAYWRIGHT: 'true',
      },
      stdout: 'pipe',
    },
    {
      command: 'pnpm website start',
      env: {
        PLAYWRIGHT: 'true',
      },
      port: 3000,
      stdout: 'pipe',
    },
    {
      command: 'pnpm api build',
      env: {
        PLAYWRIGHT: 'true',
      },
      stdout: 'pipe',
    },
    {
      command: 'pnpm api start',
      env: {
        PLAYWRIGHT: 'true',
      },
      port: 3010,
      stdout: 'pipe',
    },
  ],
```

One nit / discomfort / DX downside about this: semantically the configuration feels a bit weird, because the build command is semantically not a web server.

**Comment by karlhorky** [comment]
Oh and now we have duplicated the `PLAYWRIGHT: 'true'` line 4 times 😮

So that would be even nicer now if Playwright would set this by default.

**Comment by Skn0tt** [comment]
The webserver commands are spawned sequentially from top to bottom, but we don't block until they're ready currently. So if you depend on them running one after another, that won't work.

Here's another idea for solving your problem, would that work?

```ts
const config: PlaywrightTestConfig = {
  webServer: [ 
    {
      command: 'pnpm website build && pnpm website start',
      env: {
        PLAYWRIGHT: 'true',
      },
      port: 3000,
      stdout: 'pipe',
    },
    {
      command: 'pnpm api build && pnpm api start',
      env: {
        PLAYWRIGHT: 'true',
      },
      port: 3010,
      stdout: 'pipe',
    },
  ],
```

> Oh and now we have duplicated the PLAYWRIGHT: 'true' line 4 times 😮

Well, the alternative to this unsettling discomfort of a duplicated environment variable is 🪄 unreliable magic 🪄, as we already found out :D

Also keep in mind: If your builds need to know that they're being tested, that can be a sign that they deviate too much from the production builds.

**Comment by karlhorky** [comment]
> unreliable magic

I think it's not too magical / unreliable - it's a pretty common pattern across the testing ecosystem for tools to set environment variables. Not only Vite but also other testing tooling uses similar conventions. (not to mention production environment variables, which are almost always set by providers)

Is the Playwright team against setting `PLAYWRIGHT='true'` by default for `webServer`s?

> If your builds need to know that they're being tested, that can be a sign that they deviate too much from the production builds.

Agreed. There's a balance though - some things such as [the reCAPTCHA example in the issue description](https://github.com/microsoft/playwright/issues/32356#:~:text=features%20which%20do%20not%20work%20in%20Playwright%20(eg.%20reCAPTCHA%20or%20other%20connections%20to%20external%20services)) are valid use cases of environment variables for modifying builds and runtime environments. Playwright making this workflow easier for users would be really helpful.

**Comment by karlhorky** [comment]
> The big difference is that other tools like Vitest or Jest or anything else all control the build process, so it's a lot easier for them to reliably inject an environment variable

> We could easily inject an env var into `webServer`, but not into build scripts inside `package.json` for example. Both are valid ways of building the app though, and you'd expect any kind of "magic" we do to work the same across both.

Agreed that there may be that expectation, in case teams have lacking understanding of:

1. How environment variables integrate with their build process, especially with front-end code
2. That their build process has traditionally run outside of Playwright's purview (maybe this changes, if the build commands feature linked below is accepted)

> > Separate webServer build commands (maybe I should open a separate feature request issue?)
>
> Yes please, a separate feature request would be wonderful!

Opened an issue for `webServer` configuration for build commands here:

- https://github.com/microsoft/playwright/issues/32371

**Comment by karlhorky** [comment]
Oh, I found a further place that I would suggest this being set, if possible:

- In test code eg. `playwright/abc.spec.ts`

This case would be useful in the situation where Playwright tests import from a file type which is unsupported by Playwright's loaders config (eg. MDX files containing fixture data useful for the Playwright test) - [Playwright's loaders do not currently support extending to other file types](https://github.com/microsoft/playwright/issues/26822). Reading `process.env.PLAYWRIGHT` allows us to branch and provide a different set of data for Playwright, instead of rearchitecting file structure and imports for Playwright tests or hardcoding fixture data in Playwright tests.

I've added it to the issue description above.

## Workaround

Set `process.env.PLAYWRIGHT = 'true';` in `playwright.config.ts`:

`playwright.config.ts`

```ts
import { devices, PlaywrightTestConfig } from '@playwright/test';

process.env.PLAYWRIGHT = 'true'

const config: PlaywrightTestConfig = {
  webServer: [
    {
      command: 'pnpm website start',
      env: {
        PLAYWRIGHT: process.env.PLAYWRIGHT,
      },
      port: 3000,
      stdout: 'pipe',
    },
    {
      command: 'pnpm api start',
      env: {
        PLAYWRIGHT: process.env.PLAYWRIGHT,
      },
      port: 3010,
      stdout: 'pipe',
    },
  ],
```

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you disagree with this closure, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.
